text/0095-cognito-construct-library/working-backwards-readme.md

```diff
@@ -0,0 +1,567 @@
+
+> Internal Note: IdentityPool-SupportedLoginProviders takes a JSON with all of the providers configured.
+
+Besides the external identity providers, Cognito identity pools also support developer authenticated identities. Read
+[the documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/developer-authenticated-identities.html)
+to learn more about developer authenticated identities. This can be configured via the property `developerProvider`.
+
+Code snippet with all of the external providers configured -
+
+```ts
+new IdentityPool(this, 'myidentitypool', {
+  // ...
+  identityProviders: {
+    amazon: { appId: 'amzn1.application.188a56d827a7d6555a8b67a5d' },
+    facebook: { appId: '7346241598935555' },
+    google: { clientId: '123456789012.apps.googleusercontent.com' },
+    twitter: { consumerKey: 'xvz1evFS4wEEPTGEFPHBog', consumerSecret: 'kAcSOqF21Fu85e7zjz7ZN2U4ZRhfV3WpwPAoE3Z7kBw' },
+    developerProvider: 'login.mycompany.myapp'
+  }
+});
+```
+
+### Access Control
+
+Identity pools are configured with IAM roles from which temporary credentials will be provided to users of this pool
+when they need to access AWS resources. The IAM roles are configured with permissions based on what the users are
+allowed to access.
+
+An identity pool can have separate IAM roles configured for its authenticated users (authenticated via one of the
+identity providers as above) and unauthenticated users. They can be specified via the properties `authenticatedRole` and
+`unauthenticatedRole` properties. When no role is specified against these properties, CDK will create a new role with
+the permissions to Amazon Mobile Analytics and Amazon Cognito Sync. This is similar to the behaviour of the AWS console.
+Read the [IAM Roles documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/iam-roles.html) to learn
+more on how to set up your own IAM roles.
+
+```ts
+const authenRole = new iam.Role(this, 'cognito-authenticated-role', { ... });
+
+const unAuthenRole = new iam.Role(this, 'cognito-unauthenticated-role', { ... });
+
+new IdentityPool(this, 'my-awesome-id-pool', {
+  // ...
+  // ...
+  authenticatedRole: authenRole,
+  unAuthenticatedRole: unauthenRole
+});
+```
+
+The access control policies for a user can be extended so the user can obtain even more permissions, when they match
+specific rules. This can be specified via the `roleMappings` property. Learn more about [using rules to assign roles to
+users](https://docs.aws.amazon.com/cognito/latest/developerguide/role-based-access-control.html#using-rules-to-assign-roles-to-users).
+
+> Is there a better name than `roleMappings`?
+
+The following code snippet sets up two rules that assign the admin role based on whether the admin logged in via
+facebook or via google.
+
+```ts
+const adminRole = new iam.Role(this, 'admin-role', { ... });
+
+new IdentityPool(this, 'my-awesome-id-pool', {
+  // ...
+  // ...
+  roleMappings: {
+    'facebook': new FacebookRoleMapping({
```

Comment:
Can't the key be implicit in the mapping class?

Reply:
Also, doesn't it make sense to push these mappings into the Facebook provider?

```ts
new IdentityPool(this, 'myidentitypool', {
  // ...
  identityProviders: [
    IdentityProvider.Facebook('7346241598935555' , {
      roleMappingRules: [
        {
          // if the identifier of the user as returned by facebook is 'boris'
          claim: FacebookRoleMappingClaim.SUB,
          matchType: MatchType.EQUALS,
          value: 'boris',
          role: adminRole
        }
      ],
    })
 ]
});
```

Reply:
Nice! I like this.